ffmpeg: update to 3.2.3, add libshine support and cleanups

antonlacon opened the pull request:
Maintainer: @thess 
Compile tested: mips/ar71xx OpenWrt trunk
Run tested: None

Description:
Updates ffmpeg to 3.2.3 and wires in support for libshine.

Also makes the following changes:

Updates project URL and moves libpthread to a common dependency;
Makes libopus support in libffmpeg-full contingent on selecting the libopus package, like other external libraries;
Adds note regarding disabling altivec in ppc hard-float;
Reorganizes SIMD sections to be together;
Drops the need to specify decoder=libopus from -full, done by default;
Adds encoder=libopus to custom.

Expand options for libffmpeg-custom:
	Build licensing: GPL, GPLv3, LGPLv3
	Building for performance or size

Drops version 3 of GPL for -audio-dec, -mini, and -full. Version 3 is at the discretion of the packager, per:
	http://ffmpeg.org/doxygen/trunk/md_LICENSE.html

This contains some updates relating to issue #3914. More changes will be necessary to resolve that issue, but these should be non-controversial ones. I'll do run testing later in the week, after getting a LEDE environment up, but I suspect others will want to test this too.

thess commented:
Give me a couple of days to review this - In general I like this direction.
IIRC there were some runtime problems when enabling SSE for Geode and i486.  (See libx264 and opus CPU blacklist logic)

diizzyy commented:
SSE etc should only be enabled on 64-bit.

thess commented:
@antonlacon - I merged your work here with some other changes I was going to include in this update. In addition to your changes, I've added a new -nonfree package which by default builds with all selected external libraries (`x264`, `lame-lib` & `fdk-aac`) if BUILD_PATENTED is selected. The -full variant is the same except for these libs. If BUILD_PATENTED is not selected, the -nonfree variant is not built. 

Anyone wanting to check out the proposed configs can try:

`git pull https://github.com/thess/OpenWrt-packages.git ffmpeg-work`

antonlacon commented:
The -nonfree is going to rankle the ffmpeg folk. Mixing enable-gpl, enable-nonfree and then distributing it is what caused issues for Handbrake (and likely why it isn't packaged by many distributions). If pursuing this, a better approach might be to split on -full-gpl, and -full-lgpl.

-full-gpl would have libx264 and use ffmpeg's native aac encoder. FFmpeg believes their native aac encoder is competitive with fdk_aac at 128kbps: http://ffmpeg.org/#aac_encoder_stable

-full-lgpl would have fdk_aac but no libx264. This should be compatible for distribution per: http://ffmpeg.org/doxygen/trunk/md_LICENSE.html (see incompatible libraries' FDK section at the bottom).

With quality being near equal, fdk_aac's performance advantage is likely because of it being an integer encoder. I believe that any platform where using libx264 is worth the electricity would have an FPU, making that less of an advantage. This would still open up aac encoding at a reasonable performance for the soft-float platforms.

Mixing enable-gpl and nonfree should only be possible with -custom, as the project neither builds it nor distributes it.

As to the patch on your branch:

For -custom, the enable-encoder=libshine,libopus and decoder=libopus portions of those lines should all be there. It's more likely that the x264 and mp3-lame lines are wrong than these ones are. --disable-everything previously in the config setup is what wipes out their support and why they need to be respecified. You can verify this with the end of the ./configure output listing supported encoders/decoders. -full doesn't carry the disable-everything, which is why this extra specification isn't necessary.

The @BUILD_PATENTED flag should be on all of the libffmpeg variants except -custom. -custom handles it inside Config.in. The enable-nonfree switch is not a switch for patent encumbered code; it's a switch for saying complex licensing is involved, possibly making it non-redistributable. I know the others have a patent problem because the mp3 patents alone don't full expire until the end of the year, if Wikipedia is to be believed.

diizzyy commented:
I think you're reading a bit too much into that as far as sound quality goes.... 
https://trac.ffmpeg.org/wiki/Encode/AAC
http://wiki.hydrogenaud.io/index.php?title=AAC_encoders
Hydrogenaudio forum users also do tests rather frequently themselves too so I'd say that both are accurate in that regard but the non-free is indeed troublesome.

Why can't we ship libshine on non fpu platforms and lame on those with fpu?

thess commented:
No special order...

* The -gpl / -lgpl split makes more sense to me than what I proposed.
* Agree with the extra enable-encoder/decoder options - I missed that in the custom variant.
* OK with BUILD_PATENTED on all variants I guess. There are other packages which will not build either due to their libffmpeg dependencies.
* As far as MP3 encoding  - I like the suggestion to use shine vs lame based on HARD_FLOAT.
* Since fdk-aac licensing is problematic, the most cautious plan would be to make available only in a custom build.

There is no doubt that fdk-aac is faster. Other than license issues, the size cost is ~300K. Likewise the size cost of x264 is ~420K with it's own licensing issues.

@antonlacon - Do you want to revise this PR or should I do another pass on mine? Up to you...

antonlacon commented:
I'll revise it. I should have time over the weekend.

antonlacon commented:
Added a second patch with my changes over the weekend. 

I've addressed the -custom and @build_patented issues above, and integrated the fixes I spotted in thess's branch.

I switched libffmpeg-mini over to the lgpl licensing, and I suspect that -audio-dec could be as well, but left it alone.

After splitting libshine/lame based on hard/soft float, I wasn't sure it was worth building out another full variant for lgpl or gpl versions. Instead, I've split the full package into lgpl and gpl based on the usage of soft-float. My thought process is that a soft-float platform isn't going to do video manipulation, which dropped the need for libpostproc and the video filters. This also only grabs libopus and libshine for external libraries. The hard-float version of libffmpeg-full uses libopus, lame-lib, and libx264, and is licensed under the gpl. 

Additionally, I've framed the libfdk-aac support for libffmpeg-custom, though it has an issue while packaging.

```
Package libffmpeg-custom is missing dependencies for the following libraries:
libfdk-aac.so.1
```

I'm not sure what it's unhappy about, as I've named the fdk-aac package as a dependency, and I can see said package is including libfdk-aac.so.1. I've commented out its selection in Config.in for now.

I figured someone else might want to take a look at it at this point.

thess commented:
Many thanks for this work - I'll pull it together, do a little touch-up and merge it in the next day or 2.

thess commented:
Merged this work as: edc982430cc93144b0e6edb9fc49f04275eb69ed

I fixed the fdk-aac dependency issue, updated to 3.2.4 and did a little Makefile simplification.
Once again Ian, thanks for your work.

Also closes #3914 and #3830 (for now at least)
